pkg/discovery/vm_builder.go

```
@@ -0,0 +1,151 @@
+	// is not deleted during reconciliation
+	// TODO: create a builder for proxy VMs
+	ipProp := &proto.EntityDTO_EntityProperty{
+		Namespace: &DEFAULT_NAMESPACE,
+		Name:      &ipPropName,
+		Value:     &ipAddress,
+	}
+	entityDTOBuilder = entityDTOBuilder.WithProperty(ipProp)
+	glog.Infof("[NodeBuilder] Parse node: The ip of vm to be reconcile with is %s", ipAddress)
+	metaData := generateReconciliationMetaData()
+
+	glog.Info("%s: metadata %s", agentInfo.IP, metaData)
+	entityDTOBuilder = entityDTOBuilder.ReplacedBy(metaData)
+	entityDto, _ := entityDTOBuilder.Create()
+	return entityDto, nil
+}
+
+func generateReconciliationMetaData() *proto.EntityDTO_ReplacementEntityMetaData {
+	replacementEntityMetaDataBuilder := builder.NewReplacementEntityMetaDataBuilder()
+	replacementEntityMetaDataBuilder.Matching(PROXY_VM_IP)
+	replacementEntityMetaDataBuilder.
+	PatchSelling(proto.CommodityDTO_CPU_PROVISIONED).
+		PatchSelling(proto.CommodityDTO_MEM_PROVISIONED).
+		PatchSelling(proto.CommodityDTO_CLUSTER).
+		PatchSelling(proto.CommodityDTO_VCPU).
+		PatchSelling(proto.CommodityDTO_VMEM).
+		PatchSelling(proto.CommodityDTO_VMPM_ACCESS)
+	metaData := replacementEntityMetaDataBuilder.Build()
+	return metaData
+}
+
+// ========================== Metrics and commodities =========================================
+func (nb *VMEntityBuilder) vmCommSold(agentInfo *data.Agent, nodeMetrics *data.EntityResourceMetrics) ([]*proto.CommodityDTO, error) {
+	if agentInfo == nil {
+		return nil, fmt.Errorf("[VMEntityBuilder] Null agent while creating commodities")
+	}
+	if nodeMetrics == nil {
+		return nil, fmt.Errorf("[VMEntityBuilder] Null node metrics for agent %s", agentInfo.Id)
+	}
+	var commoditiesSold []*proto.CommodityDTO
+	// MemProv
+	memProvCap := getValue(nodeMetrics, data.MEM_PROV, data.CAP)
+	memProvUsed := getValue(nodeMetrics, data.MEM_PROV, data.USED)
+
+	memProvComm, _ := builder.NewCommodityDTOBuilder(proto.CommodityDTO_MEM_PROVISIONED).
```

do we want to ignore the error?